
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
      <link rel="icon" href="../assets/images/SIB_logo.svg">
      <meta name="generator" content="mkdocs-1.1.2, mkdocs-material-7.1.6">
    
    
      
        <title>Exercises - Spatial transcriptomics workshop</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.875de78c.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.f1a3b89f.min.css">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>:root{--md-text-font-family:"Roboto";--md-code-font-family:"Roboto Mono"}</style>
      
    
    
    
      <link rel="stylesheet" href="../stylesheets/extra.css">
    
    
      
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="none" data-md-color-accent="none">
  
    
    <script>function __prefix(e){return new URL("..",location).pathname+"."+e}function __get(e,t=localStorage){return JSON.parse(t.getItem(__prefix(e)))}</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#source-of-data" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Spatial transcriptomics workshop" class="md-header__button md-logo" aria-label="Spatial transcriptomics workshop" data-md-component="logo">
      
  <img src="../assets/images/SIB_logo.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Spatial transcriptomics workshop
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Exercises
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" data-md-state="active" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        
<a href="https://github.com/sib-swiss/spatial-transcriptomics/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/spatial-transcriptomics
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Spatial transcriptomics workshop" class="md-nav__button md-logo" aria-label="Spatial transcriptomics workshop" data-md-component="logo">
      
  <img src="../assets/images/SIB_logo.svg" alt="logo">

    </a>
    Spatial transcriptomics workshop
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/sib-swiss/spatial-transcriptomics/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/spatial-transcriptomics
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../precourse/" class="md-nav__link">
        Precourse preparations
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../course_schedule/" class="md-nav__link">
        Course schedule
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../materials/" class="md-nav__link">
        Materials
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Exercises
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Exercises
      </a>
      
        
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#source-of-data" class="md-nav__link">
    Source of data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#import-pre-processed-chromium-data" class="md-nav__link">
    Import pre-processed Chromium data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#visium-analysis" class="md-nav__link">
    Visium analysis
  </a>
  
    <nav class="md-nav" aria-label="Visium analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#exercise-the-last-one-" class="md-nav__link">
    Exercise  (the last one  ) -
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../links/" class="md-nav__link">
        Useful links
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#source-of-data" class="md-nav__link">
    Source of data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#import-pre-processed-chromium-data" class="md-nav__link">
    Import pre-processed Chromium data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#visium-analysis" class="md-nav__link">
    Visium analysis
  </a>
  
    <nav class="md-nav" aria-label="Visium analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#exercise-the-last-one-" class="md-nav__link">
    Exercise  (the last one  ) -
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/sib-swiss/spatial-transcriptomics/edit/master/docs/exercises.md" title="Edit this page" class="md-content__button md-icon">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
                  </a>
                
                
                <p>In this section, you will find the R code that we will use during the course. We will discuss the code and output during practical sessions.</p>
<h2 id="source-of-data">Source of data</h2>
<p>We will work with spatial transcriptomics data of human breast tumor, that was profiled using 3 10x technologies: Chromium scRNAseq on FFPE-fixed single cells, Visium on FFPE slide, and Xenium  <a href="https://www.biorxiv.org/content/10.1101/2022.10.06.510405v1">Janesick et al, bioRxiv</a>. </p>
<p>Before starting the exercises, set the working directory to where you have downloaded and unzipped the <a href="">data folder</a> with the files for the exercises and load the necessary packages. Make sure you download also the R script with the <a href="">custom functions</a> that will be sourced.
On the materials page, you may download an Rmd file that contains the code presented here.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>When using setwd(), change the path within quotes to where you have saved the data for the exercises</p>
</div>
<div class="highlight"><pre><span></span><code><span class="c1"># Change the path here:</span>
<span class="nf">setwd</span><span class="p">(</span><span class="s">&quot;/path/to/whereDataFolderIsSaved/&quot;</span><span class="p">)</span>

<span class="nf">library</span><span class="p">(</span><span class="n">here</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">Seurat</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">BayesSpace</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">patchwork</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">tibble</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">scater</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">grid</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">hrbrthemes</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">gridExtra</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">RColorBrewer</span><span class="p">)</span>

<span class="n">myPalette</span> <span class="o">&lt;-</span> <span class="nf">colorRampPalette</span><span class="p">(</span><span class="nf">rev</span><span class="p">(</span><span class="nf">brewer.pal</span><span class="p">(</span><span class="m">11</span><span class="p">,</span> <span class="s">&quot;Spectral&quot;</span><span class="p">)))</span>
<span class="nf">source</span><span class="p">(</span><span class="s">&quot;utils_new.R&quot;</span><span class="p">)</span>
</code></pre></div>
<h2 id="import-pre-processed-chromium-data">Import pre-processed Chromium data</h2>
<p>We will generate a Seurat object containing the Chromium scRNAseq data. The annotation of the cell types was performed by 10X using marker genes and a cell atlas reference. For the sake of time, we will directly load the saved Chromium data.</p>
<p><div class="highlight"><pre><span></span><code><span class="c1"># chrom_path &lt;- &quot;path/to/where/filtered_feature_bc_matrix.h5/isSaved&quot;</span>

<span class="nf">if</span><span class="p">(</span><span class="kc">FALSE</span><span class="p">)</span> <span class="p">{</span>
<span class="n">Counts</span> <span class="o">&lt;-</span> <span class="n">Seurat</span><span class="o">::</span><span class="nf">Read10X_h5</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="n">chrom_path</span><span class="p">,</span> <span class="s">&quot;filtered_feature_bc_matrix.h5&quot;</span><span class="p">))</span> <span class="c1"># # 18082 30365 barcodes</span>

<span class="n">MT</span> <span class="o">&lt;-</span> <span class="nf">read.csv</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="n">chrom_path</span><span class="p">,</span> <span class="s">&quot;FFPE_cell_annotations.csv&quot;</span><span class="p">))</span> <span class="c1"># 27472 barcodes</span>
<span class="n">MT</span> <span class="o">&lt;-</span> <span class="nf">column_to_rownames</span><span class="p">(</span><span class="n">MT</span><span class="p">,</span> <span class="s">&quot;Barcode&quot;</span><span class="p">)</span>

<span class="n">overlap_cell_barcode</span> <span class="o">&lt;-</span> <span class="nf">intersect</span><span class="p">(</span><span class="nf">rownames</span><span class="p">(</span><span class="n">MT</span><span class="p">),</span> <span class="nf">colnames</span><span class="p">(</span><span class="n">Counts</span><span class="p">))</span> <span class="c1"># 27460</span>
<span class="nf">length</span><span class="p">(</span><span class="n">overlap_cell_barcode</span><span class="p">)</span>

<span class="n">Counts_</span> <span class="o">&lt;-</span> <span class="n">Counts</span><span class="p">[,</span> <span class="nf">colnames</span><span class="p">(</span><span class="n">Counts</span><span class="p">)</span> <span class="o">%in%</span> <span class="n">overlap_cell_barcode</span><span class="p">]</span> <span class="c1"># 18082 27460</span>

<span class="n">MT_</span> <span class="o">&lt;-</span> <span class="n">MT</span> <span class="o">%&gt;%</span>
  <span class="nf">filter</span><span class="p">(</span><span class="nf">rownames</span><span class="p">(</span><span class="n">MT</span><span class="p">)</span> <span class="o">%in%</span> <span class="n">overlap_cell_barcode</span><span class="p">)</span> <span class="c1"># 27460 1</span>

<span class="n">chrom</span> <span class="o">&lt;-</span> <span class="nf">CreateSeuratObject</span><span class="p">(</span><span class="n">counts</span> <span class="o">=</span> <span class="n">Counts_</span><span class="p">,</span>
                          <span class="n">meta.data</span> <span class="o">=</span> <span class="n">MT_</span><span class="p">)</span>

<span class="nf">saveRDS</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="s">&quot;./data/chrom_raw.rds&quot;</span><span class="p">)</span>
<span class="p">}</span> <span class="n">else</span> <span class="p">{</span>
<span class="n">chrom</span> <span class="o">&lt;-</span> <span class="nf">readRDS</span><span class="p">(</span><span class="s">&quot;data/chrom_raw.rds&quot;</span><span class="p">)</span>
<span class="p">}</span>
</code></pre></div>
We generate some QC plots of the scRNAseq data.</p>
<p><div class="highlight"><pre><span></span><code><span class="n">chrom</span> <span class="o">&lt;-</span> <span class="nf">readRDS</span><span class="p">(</span><span class="s">&quot;data/chrom_raw.rds&quot;</span><span class="p">)</span>
<span class="n">chrom</span> <span class="o">&lt;-</span> <span class="nf">preprocess_SEU</span><span class="p">(</span><span class="n">chrom</span><span class="p">)</span>

<span class="n">chrom_libsize_drop</span> <span class="o">&lt;-</span> <span class="n">chrom</span><span class="o">$</span><span class="n">libsize_drop</span>
<span class="n">chrom_mito_drop</span> <span class="o">&lt;-</span> <span class="n">chrom</span><span class="o">$</span><span class="n">mito_drop</span>
<span class="n">chrom_lowgenecount_drop</span> <span class="o">&lt;-</span> <span class="nf">unlist</span><span class="p">(</span><span class="n">chrom</span><span class="p">[[</span><span class="s">&quot;RNA&quot;</span><span class="p">]][[</span><span class="s">&quot;lowgenecount_drop&quot;</span><span class="p">]])</span>

<span class="nf">if</span><span class="p">(</span><span class="nf">any</span><span class="p">(</span><span class="n">chrom_libsize_drop</span><span class="p">)){</span>
  <span class="n">p1</span> <span class="o">&lt;-</span> <span class="nf">plot_Hist_Low_Lib_Sizes</span><span class="p">(</span><span class="n">chrom</span><span class="p">)</span>
<span class="p">}</span> <span class="n">else</span><span class="p">{</span>
  <span class="nf">print</span><span class="p">(</span><span class="s">&quot;No Chromium cells with low library size detected by scuttle::isOutlier()&quot;</span><span class="p">)</span>
<span class="p">}</span>

<span class="nf">if</span><span class="p">(</span><span class="nf">any</span><span class="p">(</span><span class="n">chrom_mito_drop</span><span class="p">)){</span>
  <span class="n">p2</span> <span class="o">&lt;-</span> <span class="nf">plot_Hist_High_Mito_Props</span><span class="p">(</span><span class="n">chrom</span><span class="p">)</span> <span class="c1"># TODO: check plot label</span>
<span class="p">}</span>

<span class="nf">if</span><span class="p">(</span><span class="nf">any</span><span class="p">(</span><span class="n">chrom_lowgenecount_drop</span><span class="p">)){</span>
  <span class="n">p3</span> <span class="o">&lt;-</span> <span class="nf">plot_Hist_Low_Abun_Genes</span><span class="p">(</span><span class="n">chrom</span><span class="p">)</span>
<span class="p">}</span>
</code></pre></div>
We retain the cells that pass QC, and save the Chromium reference for later use with cell-type deconvolution per spot.</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Subset</span>
<span class="n">chrom</span> <span class="o">&lt;-</span> <span class="n">chrom</span><span class="p">[</span><span class="o">!</span><span class="n">chrom_lowgenecount_drop</span><span class="p">,</span> <span class="o">!</span><span class="n">chrom_libsize_drop</span> <span class="o">&amp;</span> <span class="o">!</span><span class="n">chrom_mito_drop</span><span class="p">]</span>

<span class="nf">dim</span><span class="p">(</span><span class="n">chrom</span><span class="p">)</span> <span class="c1"># 14026 25817</span>

<span class="c1"># saveRDS(chrom, &quot;data/chrom_qcd.rds&quot;)</span>
</code></pre></div>
<h1 id="visium-analysis">Visium analysis</h1>
<p>We import a Visium Seurat object, and visualize the per spot total count in 2D coordinates.</p>
<div class="highlight"><pre><span></span><code><span class="n">vis_path</span> <span class="o">&lt;-</span> <span class="s">&quot;path/to/Visium_FFPEfiles&quot;</span>

<span class="o">?</span><span class="n">Load10X_spatial</span>

<span class="n">vis</span> <span class="o">&lt;-</span> <span class="nf">Load10X_Spatial</span><span class="p">(</span><span class="n">vis_path</span><span class="p">)</span>
<span class="nf">Idents</span><span class="p">(</span><span class="n">vis</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="s">&quot;&quot;</span>
<span class="nf">SpatialFeaturePlot</span><span class="p">(</span><span class="n">vis</span><span class="p">,</span> <span class="n">features</span> <span class="o">=</span> <span class="s">&quot;nCount_Spatial&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="nf">theme</span><span class="p">(</span><span class="n">legend.position</span> <span class="o">=</span> <span class="s">&quot;right&quot;</span><span class="p">)</span>

<span class="c1"># saveRDS(vis, &quot;data/vis_raw.rds&quot;)</span>
</code></pre></div>
<h2 id="exercise-the-last-one-">Exercise  (the last one <img alt="🌞" class="twemoji" src="https://twemoji.maxcdn.com/v/latest/svg/1f31e.svg" title=":sun_with_face:" /> <img alt="🧑🏽‍🔬" class="twemoji" src="https://twemoji.maxcdn.com/v/latest/svg/1f9d1-1f3fd-200d-1f52c.svg" title=":scientist_tone3:" />) -</h2>
                
              
              
                


              
            </article>
          </div>
        </div>
        
      </main>
      
        
<footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        <a href="../materials/" class="md-footer__link md-footer__link--prev" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Materials
            </div>
          </div>
        </a>
      
      
        <a href="../links/" class="md-footer__link md-footer__link--next" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Useful links
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
        
      </div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "..", "features": [], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing"}, "search": "../assets/javascripts/workers/search.d351de03.min.js", "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.34eae1b6.min.js"></script>
      
    
  </body>
</html>