
**limma** for differential gene expression analysis of microarray or RNA seq data, also includes functions for enrichment analysis

[https://www.bioconductor.org/packages/devel/bioc/vignettes/limma/inst/doc/usersguide.pdf](https://www.bioconductor.org/packages/devel/bioc/vignettes/limma/inst/doc/usersguide.pdf)

**edgeR** for differential gene expression analysis of RNA seq data

[https://www.bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf](https://www.bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf)

**DESeq2** for differential gene expression analysis of RNA seq data

[http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)

**From reads to genes to pathways** Pipeline for DGE analysis and pathway analysis

[https://f1000research.com/articles/5-1438](https://f1000research.com/articles/5-1438) 

**clusterProfiler** e-Book

[https://yulab-smu.github.io/clusterProfiler-book/](https://yulab-smu.github.io/clusterProfiler-book/)

**fgsea** package with fgsea() function called by clusterProfiler

[https://bioconductor.org/packages/release/bioc/html/fgsea.html](https://bioconductor.org/packages/release/bioc/html/fgsea.html)

**GOsummaries** bioconductor package, to visualise Gene Ontology (GO) enrichment analysis results on gene lists arising from different analyses such clustering or PCA. The significant GO categories are visualised as word clouds that can be combined with different plots summarising the underlying data.

[https://www.bioconductor.org/packages/release/bioc/html/GOsummaries.html](https://www.bioconductor.org/packages/release/bioc/html/GOsummaries.html)

**bioconductor** Introduction and structure

[https://ivanek.github.io/analysisOfGenomicsDataWithR/02_IntroToBioc_html.html](https://ivanek.github.io/analysisOfGenomicsDataWithR/02_IntroToBioc_html.html)

**online tool** for overrepresentation analysis

[http://www.pantherdb.org/](http://www.pantherdb.org/)

**online tool** for gene set enrichment analysis

[http://www.webgestalt.org/](http://www.webgestalt.org/)

**Revigo and rrvgo**

[http://revigo.irb.hr/](http://revigo.irb.hr/)

[http://bioconductor.org/packages/release/bioc/html/rrvgo.html](http://bioconductor.org/packages/release/bioc/html/rrvgo.html)


A [recent publication](https://academic.oup.com/nar/article/49/D1/D325/6027811) of the **Gene Ontology** consortium. They try to improve taxon constraints, i.e. to better define
which terms are relevant for some species but not for others (such as leukocyte-related terms that should be constrained to vertebrates.

A [recent publication (2020)](https://www.frontiersin.org/articles/10.3389/fgene.2020.00654/full) on gene set analysis challenges, opportunities and future research. 

**Tools for species other than human or mouse**.
Bioconductor contains "org.db" packages with genome/gene annotation for many organisms, for example
for [Drosophila](https://bioconductor.org/packages/release/data/annotation/html/org.Dm.eg.db.html) or [Arabidopsis](https://bioconductor.org/packages/release/data/annotation/html/org.At.tair.db.html).

For fungi, [FungiDB](https://fungidb.org/fungidb/app/search/pathway/PathwaysByGeneList) provides identification of metabolic pathways based on list of genes

PlantGSAD: a gene set database for plants [https://academic.oup.com/nar/article/50/D1/D1456/6371972?login=true](https://academic.oup.com/nar/article/50/D1/D1456/6371972?login=true)

Eventually this repository by [Zhang et al](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-020-3408-y)

